Add cache-group input, eslint, remove deploy_sha output

nonrational opened the pull request:
It would be handy to be able to cache distinct results inside a single SHA. This PR adds the ability to specify a `cache-group` which will get added to the cache key. 

I had some trouble getting the `ci.yml` to run against the _current_ sha. I couldn't get `uses: owner/repo@${{ github.sha }}` to work, unfortunately. Accordingly, before this is ready to merge, we'll need to revert some of the changes to ci.yml and document the new input usage in the README.

icapitalan committed:
Support an optional key-prefix to support caching discrete actions for the same sha

icapitalan committed:
Update commit in ci.yml to exercise code in PR

bmuller commented:
Should we use `github.workflow` (the workflow name) as a part of the cache key?  And would that be enough to solve the issue?

nonrational commented:
> Should we use `github.workflow` (the workflow name) as a part of the cache key? And would that be enough to solve the issue?

Ah, great question. I like the idea of no-config-necessary workflow isolation. But, I appreciate the flexibility of being able to define your own cache groups, even in the same workflow. I got thinking about this because of our `mix xref` check that doesn't _have_ to break the build. I was toying with the idea of caching "low priority" jobs independently from the critical jobs. It would be annoying to break them out into a separate workflow because it'd be another file to synchronize environment and runtime versions across.

What if we called the input `cache-group` and defaulted it to `github.workflow`? That way, you'd get workflow isolation by default, but would still have the option to override it with a custom key if you want more flexibility.

bmuller commented:
> What if we called the input `cache-group` and defaulted it to `github.workflow`? That way, you'd get workflow isolation by default, but would still have the option to override it with a custom key if you want more flexibility.

Sounds great to me!

icapitalan committed:
remove default branch sha, support cache-group input

icapitalan committed:
prepare

icapitalan committed:
test new feature

icapitalan committed:
remove deploy_sha, add inputs documentation

nonrational commented:
Updated in https://github.com/parallel-markets/cache-result-action/pull/2/commits/1859e4d067db992ef860cf771d1657af8b685013. Also removed the `deploy_sha` output, since it doesn't provide the desired effect.

bmuller commented on dist/index.js:
```diff
@@ -64939,20 +64939,12 @@ const run = async () => {
     const token = core.getInput('token', { required: true })
     const octokit = github.getOctokit(token)
 
-    // Get the repo object, which contains the `default_branch` property.
-    const repoResult = await octokit.request('GET /repos/{owner}/{repo}', { owner, repo })
-    const defaultBranch = repoResult.data.default_branch
-    core.info(`Using default branch '${defaultBranch}'`)
-
-    // Fetch the current SHA of the default branch
-    const ref = `heads/${defaultBranch}`
-    const refResult = await octokit.rest.git.getRef({ owner, repo, ref })
-    core.info(`Found '${defaultBranch}' with SHA ${refResult.data.object.sha}`)
-
```
With this removal, I don't think we need the `octokit` variable, which means I don't think we need a token input.

nonrational replied:
Removed and added basic eslint to catch unused vars. 👍

icapitalan committed:
more cleanup, add linting, prepare

icapitalan committed:
rename ci, fix naming, break up steps

icapitalan committed:
rename job to test since we're building too

bmuller commented on src/index.js:
```diff
@@ -11,26 +10,12 @@ const run = async () => {
-    const key = 'cache-result-action-' + sha + '-' + Math.floor(Date.now() / 1000)
-    const cacheKey = await cache.restoreCache([RESULT_PATH], key, ['cache-result-action-' + sha])
+    const cacheGroup = core.getInput('cache-group')
+    const key = ['cache-result-action', cacheGroup, sha, Math.floor(Date.now() / 1000)].filter(Boolean).join('-')
+
+    await cache.restoreCache([RESULT_PATH], key, ['cache-result-action-' + sha])
```
The third argument to `cache.restoreCache` should have the full cache path before the timestamp:

```suggestion
    const keyPrefix = `cache-result-action-${cacheGroup}-${sha}`
    const key = keyPrefix + '-' + Math.floor(Date.now() / 1000)

    await cache.restoreCache([RESULT_PATH], key, [keyPrefix])
```

icapitalan committed:
Fix call to cache.restoreCache to ensure correct restoration

Co-authored-by: Brian Muller <bmuller@parallelmarkets.com>

icapitalan committed:
dist

icapitalan committed:
update to latest
